### ISSUE TITLE
Gtk+-3.21 desktop background regressions 

### ISSUE THREAD

**Issue by raveit65** [mate-desktop/caja#604]
With gtk+-3.21 we have some regressions caused by a gtk+ change.
https://git.gnome.org/browse/gtk+/commit/?id=2c7b21718f18ab743cd79784ae28cad582f58503
With changes in caja and m-s-d this issue is particular fixed.
https://github.com/mate-desktop/caja/pull/602
https://github.com/mate-desktop/mate-settings-daemon/pull/157

My test results:
- case with compositor and desktop icons.
changing desktop backgrounds/wallpapers works but no fade effect.

- case w/o compositor and with desktop icons.
changing desktop backgrounds/wallpapers does not work but i see the desktop-fallback color from caja.css.

- case with compositor and w/o desktop icons.
Changing desktop backgrounds/wallpapers works but no fade effect.

- case w/o compositor and w/o desktop icons
Changing desktop backgrounds/wallpapers and background-fade works well. :smile:

**Comment by lukefromdc** [comment]
I had no luck so far trying to draw a fallback background using cairo on the scrolledwindow (fm-desktop-icon-view).  I tried the suggested gtk_render_background, it had exactly no effect.
I tried this code in the place of the normal fallback code that sets a style property. It built and ran but could not draw even the simple test color to the background. Had it worked that could have been replaced with image loading code:

#if GTK_CHECK_VERSION(3, 21, 0)
    gboolean
    composited = gtk_widget_is_composited (GTK_WIDGET(desktop_icon_view));
   
        if (!composited){
            GtkStyleContext *context;
            cairo_t         *cr;
            cairo_pattern_t *pattern;
            context = gtk_widget_get_style_context (GTK_WIDGET(desktop_icon_view));
            cr = cairo_create(pattern);
            cairo_set_source_rgb(cr, 0.2, 0.4, 0.6 );
            gtk_render_background(context, cr, 0 ,0, allocation.width, allocation.height );
            cairo_destroy (cr);
        }
#endif

**Comment by albertsmuktupavels** [comment]
- If you create cairo_t from uninitialized cairo_pattern_t, then what do you expect?
- cairo_create needs cairo_surface_t not cairo_pattern_t.

**Comment by lukefromdc** [comment]
Problem here is simple: I have to learn all of this as I go and getting good
documentation about how to use cairo has not been easy at all. I'll play with
these suggestions and see if I get anything.

**Comment by lukefromdc** [comment]
I just found something interesting: reverting  the exclusion of
`gdk_x11_drawable_get_xid (gtk_layout_get_bin_window (GTK_LAYOUT (widget)));`
(don't force native window for the desktop canvas)
from GTK 3.21 builds not only brought back the redraw problem as expected ,but also displayed the root window background even when compositng is not in use. That again suggests the problem with the old bg code is that GTK 3.21 combines what were previously several draw surfaces into one.

Still no luck at all with cairo or gtk_render_background, this attempt did nothing:

```
#if GTK_CHECK_VERSION(3, 21, 0)
    gboolean
    composited = gtk_widget_is_composited (GTK_WIDGET(desktop_icon_view));
   
        if (!composited){
            GdkScreen *screen;
            GtkStyleContext *context;
            cairo_t         *cr;
            cairo_surface_t *surface;
            cairo_pattern_t *pattern;
            screen = gtk_widget_get_screen (GTK_WIDGET (desktop_icon_view));
            context = gtk_widget_get_style_context (GTK_WIDGET(desktop_icon_view));
            surface = gdk_window_create_similar_surface (gdk_screen_get_root_window (screen),
                                                CAIRO_CONTENT_COLOR,
                                                allocation.width, allocation.height);
            gtk_widget_set_app_paintable(desktop_icon_view, TRUE);
            cr = cairo_create(surface);
            cairo_set_source_rgb(cr, 0.2, 0.4, 0.6 );
            cairo_fill(cr);
            gtk_render_background(context, cr, 0 ,0, allocation.width, allocation.height );
            cairo_destroy (cr);
        }
#endif
```

**Comment by albertsmuktupavels** [comment]
Can you answer question - what your code is doing? I really would suggest you to create small test app where you learn how to draw on widget.

- you create new surface, check documentation if you don't know what create_similar_surface does.
- you create new cairo_t so you can draw to surface
- you fill it with rgb color
- then you try to render background
- and finally you destroy cairo_t

What did you expect from this code? If I am reading code correctly then your code did exactly what you asked - it filled surface with color and rendered background on it. You can verify that by doing this:
cairo_surface_write_to_png (surface, "background.png");

As minimum you should see rgb color in your saved imaged unless gtk_render_background has filled/painted something over it.

Then you are leaking memory... You created surface, you also need to destroy it.

If you want to draw on widget then there is draw signal where cairo_t is already available.

**Comment by lukefromdc** [comment]
I really need someone else to work on this part of the job as I simply don't
know enough about working with cairo!  I did in fact play with code from some
simple test apps that have been published, but no attempt to insert it into 
this part of Caja has ever worked for me.

I can get by just fine without compositing support at all, but other users might
have trouble with this.

**Comment by lukefromdc** [comment]
Just ran a quick test with that PNG image code, using cairo_fill
gave a black image, with cairo_paint I got an  image filled with 
the specified color but the noncomposited desktop was still black.

**Comment by albertsmuktupavels** [comment]
Did you do it in draw signal and/or vfunction? Also why do you want to draw on fm-desktop-icon-view? I would do it on caja-desktop-window...

https://bugzilla.gnome.org/show_bug.cgi?id=769360
I created patch for nautilus for non-composited screen - it just reads background from root and draws it on nautilus window.

**Comment by lukefromdc** [comment]
I finally found something that worked: turned out the relative URLs in the themes
did not find an image to load, but an absolute URL with a full path can load an
image and put it on .caja-desktop-window with no problems at all.

Thus, a fallback background could be as simple as having a symlink whose name and
path never change to whatever the current background image is-or to a dynamically
generate image of m-s-d's  root window (via cairo) stashed somewhere in the user's 
.cache directory.

**Comment by lukefromdc** [comment]
I just found that so long as an absolute path is used in the URL, the current .caja-fallback-desktop-background selector can put a background image on the desktop with the existing code. Alternately I can move it to caja-desktop-window. We'v already got a copy of the desktop image in ~/cache/mate/background, all I need now is a way to replace the random number name with a name that never changes-and point caja.css at it.  If we want to also support color backgrounds in the noncompositing case that will need to be a different image, one made directly of the root window probably using cairo. I'll play with that tonight as well

**Comment by lukefromdc** [comment]
I was able to use cairo to create an image of the root window and save it-only to find that running gdk_cairo_set_source_window got not only the background set by m-s-d but also all other open windows including the panel and conky!  This was the result:

![background_scaled](https://cloud.githubusercontent.com/assets/14036792/17454820/433af7ac-5b72-11e6-9355-24e78bcc4ef1.png)

Here at last is proof of the root window issue with GTK 3.21

I then moved on to loading images with css set for now in my theme.  I found that only the full absolute path with no file globs suceeded in loading a background image. Thus, a user named "bob" could put a file in ~/config/gtk-3.0 containing the stanza 

```
.caja-fallback-desktop-background {
	background-image: url("/home/bob/Pictures/Background");
}
```
 name the desired image "background" and put in in his "Pictures" directory and caja would pick it up, but that same stanza would break for any other user. Thus it can't go into caja.css unless the image goes outside the user's /home directory. The /tmp directory is writable but often in RAM, OK if m-s-d is caching the image there but not for a user to have to dump it there manually.

For making caja use the same background as in the compositing case, the problem becomes one of putting the image somewhere writable with a path that never changes (/tmp maybe?) and a name that never changes. Where is the code that caches an image to~/.cache/mate/background?
I could add code to it to put a second copy in /tmp with an unchanging name, then add the necessary css to caja.css and we have image support for the noncomposting case.  Remaining issue is backgrounds without an image (color/gradient) still have to be set up in ~/config/gtk-3.0 to override caja.css.

**Comment by albertsmuktupavels** [comment]
1) https://developer.gnome.org/gdk3/stable/gdk3-Cairo-Interaction.html#gdk-cairo-set-source-window
It clearly says that window contains all its subwindows when rendering, but you are expecting only root window...

2) Why do you want to save background as image? Just to get image on disk so you are able to generate css? Background image is already available for loading, check my patch in nautilus bug. Get surface and use it in draw signal.

**Comment by lukefromdc** [comment]
Unfortunately that is NOT in the current documentation you linked to. Right now I 
am trying to split the job into segments, Remember that all of this is learn by
experimentation for me, with zero formal training. Most of what I know about 
cairo I found out by looking at simple testy programs. 

OK, given that GTK can draw the background, using GTK and css to do so has
the advantage that future GTK updates are less likely to break it, The advantage
of harvesting it from m-s-d's background as drawn is support for color and gradient
backgrounds by copying to an image to a variable, a file, whatever. Unforutunately,
 GTK's css loads from a file, a "resource," or data in the form of a quoted string that can't
include variable values,but not from a variable, Thus an image copied from m-s-d must 
either be stored as a file or drawn with lower level code which may be more likely to
break in the future.  

There is good reason I am just testing all this locally for now-I will probably get lots
of garbage code until the ideas come to my by trial and error, and by thinking about
them when on the  road. When I get something that can load the same image that
the user selected from the change background callback i'll set up a branch,

If you know of a good cairo tutorial anywhere it would be very useful, the GNOME
documentation is helpful but it is not enough.

**Comment by albertsmuktupavels** [comment]
On Sat, Aug 6, 2016 at 7:03 PM, lukefromdc <notifications@github.com> wrote:

> Unfortunately that is NOT in the current documentation you linked to.
> Right now I
>

What do you mean with current documentation? That fact is mentioned even in
3.0 documentation:
https://developer.gnome.org/gdk3/3.0/gdk3-Cairo-Interaction.html#gdk-cairo-
set-source-window


> am trying to split the job into segments, Remember that all of this is
> learn by
> experimentation for me, with zero formal training. Most of what I know
> about
> cairo I found out by looking at simple testy programs.
>
> OK, given that GTK can draw the background, using GTK and css to do so has
> the advantage that future GTK updates are less likely to break it, The
> advantage
> of harvesting it from m-s-d's background as drawn is support for color and
> gradient
> backgrounds by copying to an image to a variable, a file, whatever.
> Unforutunately,
> GTK's css loads from a file, a "resource," or data in the form of a quoted
> string that can't
> include variable values,but not from a variable, Thus an image copied from
> m-s-d must
> either be stored as a file or drawn with lower level code which may be
> more likely to
> break in the future.
>

Get current background surface with mate_bg_get_surface_from_root and paint
it in draw signal. I would say you should forget about your idea...

There is good reason I am just testing all this locally for now-I will
> probably get lots
> of garbage code until the ideas come to my by trial and error, and by
> thinking about
> them when on the road. When I get something that can load the same image
> that
> the user selected from the change background callback i'll set up a branch,
>

User selected background is already on root window...

If you know of a good cairo tutorial anywhere it would be very useful, the
> GNOME
> documentation is helpful but it is not enough.
>

Use official documentation:
- https://www.cairographics.org/manual/
- https://developer.gnome.org/gdk3/
- https://developer.gnome.org/gtk3/

**Comment by raveit65** [comment]
> Background image is already available for loading, check my patch in nautilus bug

Which nautilus bug and do you have a link?

**Comment by albertsmuktupavels** [comment]
Link was already posted...
https://bugzilla.gnome.org/show_bug.cgi?id=769360

**Comment by lukefromdc** [comment]
Thanks albertsmuktupavels, your patch just plain worked after I ported it to Caja. Hope they take it for Nautilus and Nemo could use it as well. Here's a PR for Caja:

https://github.com/mate-desktop/caja/pull/607

**Comment by raveit65** [comment]
Thanks @albertsmuktupavels , you should become a Mate developer ;)
And thanks to @lukefromdc for the PR.
Works great now except this little issues.
1. in non compositor mode abrt ( fedora's bug reporting tool) noticed a caja crash if i disable 'show-desktop-icons' and start to switch wallpapers or color backgrounds.
```
Program terminated with signal SIGBUS, Bus error.
#0  background_changed (window=window@entry=0x56424c5f4440) at caja-desktop-window.c:64
64	    if (window->details->surface) {
```
full stacktrace https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/caja-backtrace-wallpaper-switch-non-compositor
Still wondering that the code is involved if caja doesn't handle the desktop.
So switching from show-desktop-icons mode to non icons seems not clean.

2. background-fade does not work with non-compositor and show-desktop-icons mode.
I guess the code for it is simply not called.

3. Using a solid color gives us a black background, same like the thumbnail which is shown in mate-appearance-properties.
I think this is not related to the PR, as showing the thumbnail for a solid bg was a regression which was left after the port to GdkRGBA.
As the bg code is now different for non compositor mode we see here the same issue as with mate-appearance-properties.

Good is that wallpaper switching is much faster as before :)
At least all is much better than after the breakage by gtk+.

**Comment by albertsmuktupavels** [comment]
>
> Thanks @albertsmuktupavels <https://github.com/albertsmuktupavels> , you
> should become a Mate developer ;)
>
No, thanks!

> And thanks to @lukefromdc <https://github.com/lukefromdc> for the PR.
> Works great now except this little issues.
> 1. in non compositor mode abrt ( fedora's bug reporting tool) noticed a
> caja crash if i disable 'show-desktop-icons' and start to switch wallpapers
> or color backgrounds.
>
Just tried with nautilus and it does not crash, I can set composited on/off
without problems, background changing works. But nautilus desktop window
most likely these days works differently.

> Program terminated with signal SIGBUS, Bus error.
> #0  background_changed (window=window@entry=0x56424c5f4440) at caja-desktop-window.c:64
> 64      if (window->details->surface) {
>
> full stacktrace https://dl.dropboxusercontent.
> com/u/49862637/Mate-desktop/Bugs/caja-backtrace-wallpaper-sw
> itch-non-compositor
> Still wondering that the code is involved if caja doesn't handle the
> desktop.
> So switching from show-desktop-icons mode to non icons seems not clean.
>
At least in past disabling desktop icons did not disable nautilus desktop
window. It simply did not show icons I think, but I might be wrong...

>
>    1.
>
>    background-fade does not work with non-compositor and
>    show-desktop-icons mode.
>    I guess the code for it is simply not called.
>
> Currently background changing with fading works like this:
- gets current bg surface as start surface
- from settings new bg surface is prepared as end surface
- fading happens on third surface - so compositor only sees old surface
and/or directly new surface.

In short - fade still works, but no one sees it...

**Comment by raveit65** [comment]
> Just tried with nautilus and it does not crash, I can set composited on/off without problems, background changing works. But nautilus desktop window most likely these days works differently.
At least in past disabling desktop icons did not disable nautilus desktop window. It simply did not show icons I think, but I might be wrong...

I think this is different in gnome nowadays, here the bg is never draw by nautilus w/o your patch.
In Mate mate-settings-daemon draw the bg with no icons, and caja draws it if the icons are enabled.
This was was changed in gnome in early 3.0 development.
And that switch works fine with caja-1.15.2, before the redraw fixes was added.

**Comment by lukefromdc** [comment]
Just confirmed the icons off/no compositing/background change bug here.

A crash when turning off icons on the desktop will be noticed by users in threee scenarios: turning icons off while Caja is in the middle of a file copy job, turning icons on the desktop back on, or plugging in a flash drive and not having it automount until they open a Caja window(which will restart Caja). The first of those three could lead to data loss if it is not noticed and the source of the data (say, a camera card) is cleaned out witnout ever looking at where the files were copied. That would be a stupid thing to do, but I know people who do such things.  

A possible fix would be to modify the runtime selector to treat "no icons on the desktop" as the composited case unconditionally. It would seem to me the desktop window should not be drawn at all without icons on the desktop, and this code should not even be reached under those conditions. It is, so I will look into reading that background key and fixing this.

**Comment by raveit65** [comment]
For clarification, the crash happens if i switch the wallpaper/backgrounds, not during switching desktop icons off.
So, i suspect that m-s-d and caja want to do the switch, but caja crashed because m-s-d handle the desktop at this point.

**Comment by raveit65** [comment]
i forgot, if i start with that settings (w/o compositor + no desktop icons), than this crash doesn't happen, because m-s-d handle th bg and not caja.
It looks like that the fix doesn't handle the case switching bg handling back to m-s-d if turning off desktop-icons.

**Comment by lukefromdc** [comment]
Background can be changed with icons on, so if Caja is still drawing the background it should be able to handle it with icons off. If not, setting it to act as if composited when icons are disabled might be the fix

**Comment by raveit65** [comment]
In general, just in that moment you disable desktop icons, desktop handling moved from caja to m-s-d in Mate, with or w/o compositor.
This is different to gnome, as far i know the bg is draw by gnome-shell in all cases.

**Comment by lukefromdc** [comment]
I just found that reading  MATE_BG_KEY_SHOW_DESKTOP and exiting either 
caja_desktop_window_composited_changed or caja_desktop_window_draw had no effect at all. Apparently the key did not get re-read on changing this preference, because an unconditional exit from the first function will stop the crash at the expense of turning the desktop black when icons are on. I'm thinking caja_desktop_window_composited_changed only runs when composition is turned on or off and nothing is getting re-run for turning icons off until Caja is restarted without them.

Also tested with both libunique and gtkapplication to see if any difference in handling the startup and desktop had an effect, same either way.

**Comment by lukefromdc** [comment]
About m-s-d handling the background, it is now drawing the background unconditionally, Caja draws it as well when not composited because the m-s-d background is covered up with a window with no alpha channel. The problem seems to be getting caja to stop trying to handle the background when icons are turned off without compositing.

**Comment by albertsmuktupavels** [comment]
Would not it be better to find out why it is crashing?

If applying patch then it might be good idea to do it fully, no? Looking at merge proposal I see that it does not include finalize part...

In nautilus now desktop window leaves in separate process, but most likely you still have it in same process. So here is what is happening:
- you disable desktop icons with gsettings
- caja sees that it should not show desktop window and destroys it
- you change background image
- gtk sees that property on root window changes and runs all attached callbacks
- filter_func is called with already freed/invalid pointer
- crash

In short - your problem is that you don't remove filter_func when window is destroyed. Now - beer from you... :P :D

**Comment by raveit65** [comment]
To which address should i pipe a good german sixback beer ;)
Well, you're right, but i've alpha freeze in a few days for fedora and this is a important date for Mate-Compiz spin. If a desktop is in general broken a spin can be dropped.
For this reason i merged the PR.
........i'm standing always at the front line with fedora.

**Comment by lukefromdc** [comment]
I was able to fix most but not all crashes on background change:
https://github.com/mate-desktop/caja/pull/610

You can still crash it deliberately by turning icons off, then turning them back on again, then changing backgrounds, possibly due to a race condition between reading the gsettings key and changing the background but anything else works.True>false>change BG OK, false>true>change BG OK, false>true>false>BG change OK, true>false>true>bg change=still crashes.

**Comment by albertsmuktupavels** [comment]
Please re-read my last comment and also check my patch again...

**Comment by raveit65** [comment]
stacktrace with latest PR:
```
Core was generated by `caja'.
Program terminated with signal SIGBUS, Bus error.
#0  background_changed (window=window@entry=0x56424c5f4440) at caja-desktop-window.c:64
64	    if (window->details->surface) {
[Current thread is 1 (Thread 0x7f71b2e0ff40 (LWP 5041))]

Thread 1 (Thread 0x7f71b2e0ff40 (LWP 5041)):
#0  0x000056424a711e23 in background_changed (window=window@entry=0x56424c5f4440) at caja-desktop-window.c:64
        screen = 0x56424c304020 [GdkX11Screen]
#1  0x000056424a7120a0 in filter_func (xevent=<optimized out>, event=<optimized out>, window=0x56424c5f4440) at caja-desktop-window.c:89
        xev = <optimized out>
        gdkatom = <optimized out>
        event = <optimized out>
        window = 0x56424c5f4440
        xevent = <optimized out>
        xev = <optimized out>
#2  0x00007f71b06bfec1 in gdk_event_apply_filters (xevent=xevent@entry=0x7ffdc499b250, event=event@entry=0x56424c317210, window=window@entry=0x56424c30b000 [GdkX11Window]) at gdkeventsource.c:79
        filter = 0x56424c688110
        node = <optimized out>
        tmp_list = 0x56424c7f7e60 = {0x56424c688110}
        result = <optimized out>
#3  0x00007f71b06c01f9 in _gdk_x11_display_queue_events (xevent=0x7ffdc499b250, event_source=0x56424c316f80) at gdkeventsource.c:205
        dpy = 0x56424c2ebc00
        event = 0x56424c317210
        result = <optimized out>
        event_translator = 0x0
```
full version: https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/caja-backtrace-no-compositor-with-icon
Happened with ' true>false>true>bg change=still crashes'.
Seems that filter_func is the problem again. Except later mentioned  _gdk_x11_display_queue_events both stacktraces are equal.

**Comment by albertsmuktupavels** [comment]
Is my English so bad? You have applied my patch only partially. You need also nautilus_desktop_window_finalize part...

Every time desktop window is created filter function is added, but never removed.

true>false>true>bg change=still crashes':
- first desktop window exists with filter function
- first desktop window is destroyed, but filter function still exists
- "second" desktop window is created and new filter function is added.

So now you have two filter functions, first has destroyed window, second is ok. You change background and gtk+ calls both filter functions. Since desktop icons is enabled, both functions tries to get background surface.

And what happens when you access already freed / invalid memory you probably know.

**Comment by raveit65** [comment]
> Is my English so bad?

hehe, i only tested the latest PR.....and yes, i understand what you saying.
At least the stacktrace confirmed what you're saying.

**Comment by raveit65** [comment]
https://gist.github.com/raveit65/51b40a20f8836562369fc155a828edc4
This fixes the crash for me, applied w/o https://github.com/mate-desktop/caja/pull/610
@lukefromdc if it works for you, you can replace the PR with it, as it is your work.
@albertsmuktupavels Thank you for the patience with us ;)

**Comment by raveit65** [comment]
wtf, this does not work, after testing this on bare metal caja crashes again.
https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/caja-backtrace-new
```
Core was generated by `caja'.
Program terminated with signal SIGBUS, Bus error.
#0  background_changed (window=window@entry=0x55fb13d90410) at caja-desktop-window.c:64
64	    if (window->details->surface) {
[Current thread is 1 (Thread 0x7fbdc5f8cf40 (LWP 2489))]

Thread 1 (Thread 0x7fbdc5f8cf40 (LWP 2489)):
#0  0x000055fb119e4ce3 in background_changed (window=window@entry=0x55fb13d90410) at caja-desktop-window.c:64
        screen = 0x55fb13b36020 [GdkX11Screen]
#1  0x000055fb119e4f60 in filter_func (xevent=<optimized out>, event=<optimized out>, window=0x55fb13d90410) at caja-desktop-window.c:89
        xev = <optimized out>
        gdkatom = <optimized out>
        event = <optimized out>
        window = 0x55fb13d90410
        xevent = <optimized out>
        xev = <optimized out>
#2  0x00007fbdc383cec1 in gdk_event_apply_filters (xevent=xevent@entry=0x7ffd8a8eb910, event=event@entry=0x7fbd940020c0, window=window@entry=0x55fb13b3d000 [GdkX11Window]) at gdkeventsource.c:79
        filter = 0x55fb13d50220
        node = <optimized out>
        tmp_list = 0x55fb13e5a080 = {0x55fb13d50220}
        result = <optimized out>
#3  0x00007fbdc383d1f9 in gdk_event_source_translate_event (xevent=0x7ffd8a8eb910, event_source=0x55fb13b4a000) at gdkeventsource.c:205
```
But my coding knowledge is very limited, probably i made a mistake :-/

**Comment by albertsmuktupavels** [comment]
Have you turned off all warnings/errors for compiler? Does not it complain about unused function - caja_desktop_window_finalize? Looking at patch you do add function, but it is never used...

GObjectClass *object_class = G_OBJECT_CLASS (...);
object_class->finalize = caja_desktop_window_finalize;

**Comment by raveit65** [comment]
> Have you turned off all warnings/errors for compiler?

How could i fixed the tons of deprecations in in the last year?
And no, there was no warning.
But .....DOW.....that fixes the crash :)
Gist at github is updated.
@lukefromdc 
Please test

**Comment by lukefromdc** [comment]
Tested it and it works here on bare metal. Closed previous PR and added this one based on Ravit65's patch:
https://github.com/mate-desktop/caja/pull/611

**Comment by raveit65** [comment]
Just merged https://github.com/mate-desktop/caja/commit/f5164f75918a8e65ccde9565c4fd9f366064172e
I will open a new report for the last regression background-fade.
Sounds like it isn't a big deal but it will break transition backgrounds which are the default in rhel/centos. Here are different color used for night, morning, day, afternoon.
And we have the cosmos background slide show. 
.....i know Alberts explanation about.

**Comment by raveit65** [comment]
https://github.com/mate-desktop/caja/issues/612

**Comment by raveit65** [comment]
@lukefromdc @albertsmuktupavels 
I noticed that compiz-reloaded (0.8.12) is also affected by this gtk+ change, the desktop becomes unusable with compiz.
Does compiz-0.9.x work?

**Comment by raveit65** [comment]
error......i meant the redraw issue :-/

**Comment by lukefromdc** [comment]
No desktop redraw issues for me with Compiz 0.9 that I have not also had with Marco and all other window managers through the entire GTK 3.21/MATE 1.15 development cycle. Right now, the only file managers that work with ANY window manager in GTK 3.21 are caja version 1.15.4 or any Nautilus including and after 3.8. With Caja 1.15.4 compiz 0.9.13 works fine on my box, using it right now,

If you are getting redraw problems in compiz 0.8 that you do NOT get in Marco, that would indicate a compiz-specific problem. Try running nautilus on the desktop, if redrawing works your issues are probably the caja version you are using-or any version of nemo, which has yet to merge an existing PR to apply the Nautilus patch from 3.7.91 that made desktop rendering work with GTK 3.21. Yes that was fixed first in Nautilus, which is why GNOME devs never noticed a problem with GTK 3.21.3 when it came out.

Compiz 0.9 with Nemo used to give even worse desktop issues than what GTK 3.21 gave Caja, possibly because Nemo did use the transparent desktop (like nautilus or caja 1.15.4 with compositing) but still forced native window use, a combination that does not work with any WM in GTK 3.21. For unknown reasons Nemo 3.0 with GTK 3.20 and recent versions of compiz does draw the desktop without issue, I think that change was somewhere in compiz but I am not sure of this.

**Comment by raveit65** [comment]
I found a workaround :-)
Using the compiz wallpaper plugin + a killall caja fixes the issue.
So we lost only the Mate integration, reading and changing backgrounds from m-c-c.

> If you are getting redraw problems in compiz 0.8 that you do NOT get in Marco, that would indicate a compiz-specific problem.

Yes, terminal output from ccsm sounds like the gtk+ change.
```
(ccsm:29981): Gtk-CRITICAL **: gtk_cell_renderer_render: assertion 'GTK_CELL_RENDERER_GET_CLASS (cell)->render != NULL' failed
```

**Comment by lukefromdc** [comment]
Are you running these tests with caja 1.15.4? If you did and got icons on the desktop over Compiz's wallpaper you just found a substantial difference between compiz 0.8 and 0.9. In compiz 0.9 the wallpaper is rendered on top the desktop, covering up icons even if the desktop itself is transparent. If Marco is working and compiz 0.8 is not, look for a forced native window somewhere in the compiz 0.8 code, maybe that was something the Ubuntu devs took out in the refactor and rewrite to C++ for 0.9

**Comment by raveit65** [comment]
yes, test was with caja from this PR from tomorrow (for you yesterday :-)
After a 'killall caja' desktop icons are there and i can load a different wallpaper from compiz plugin too.
I will grep for a forced native window in compiz code.
Thanks for the tip :-)

**Comment by albertsmuktupavels** [comment]
On Mon, Sep 5, 2016 at 6:58 PM, raveit65 <notifications@github.com> wrote:

> @lukefromdc <https://github.com/lukefromdc> @albertsmuktupavels
> <https://github.com/albertsmuktupavels>
> I noticed that compiz-reloaded (0.8.12) is also affected by this gtk+
> change, the desktop becomes unusable with compiz.
> Does compiz-0.9.x work?
>
I think it works, at least I have not seen any problems...
